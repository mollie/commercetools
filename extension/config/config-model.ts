export interface Config {
  port?: number | string;
  mollieApiKey: string;
  ctConfig: {
    projectKey: string;
    clientId: string;
    clientSecret: string;
    authUrl: string;
    host: string;
    scopes?: string[];
  };
}
