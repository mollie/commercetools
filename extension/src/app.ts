import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
