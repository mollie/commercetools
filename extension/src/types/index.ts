// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
