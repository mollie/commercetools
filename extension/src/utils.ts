import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
