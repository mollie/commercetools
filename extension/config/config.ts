import { Config } from './config-model';

const isConfigValid = (config: Config): { valid: boolean; message: string } => {
  const { mollie, commercetools } = config;
  let message = '';

  if (!mollie?.apiKey) {
    message = message + 'No Mollie API Key found\n';
  }
  if (!commercetools?.host || !commercetools?.clientId || !commercetools?.clientSecret || !commercetools?.authUrl || !commercetools?.projectKey) {
    message = message + 'Commercetools configuration requires missing required key(s)\n';
  }
  return {
    valid: !message.length,
    message,
  };
};

export function loadConfig(ctMollieConfig: string | undefined) {
  console.info("loadConfig : " + ctMollieConfig);
  try {
    const envConfig = JSON.parse(ctMollieConfig || '');

    const config: Config = {
      ...envConfig,
      service: {
        port: envConfig.service?.port || 3000,
        logLevel: process.env.LOG_LEVEL || envConfig.service?.logLevel || 'info',
        logTransports: envConfig.service?.logTransports || 'terminal',
      },
    };

    const localeRegex = /^[a-z]{2}_[A-Z]{2}$/;

    Object.assign(
      config.service,
      envConfig.service?.webhookUrl && { webhookUrl: envConfig.service.webhookUrl },
      envConfig.service?.redirectUrl && { redirectUrl: envConfig.service.redirectUrl },
      envConfig.service?.locale && envConfig.service.locale.match(localeRegex) && { locale: envConfig.service.locale },
    );
    Object.assign(config.commercetools, envConfig.commercetools?.authentication ? { authentication: envConfig.commercetools.authentication } : { authentication: { isBasicAuth: false } });
    Object.assign(config.commercetools, { enableRetry: envConfig.commercetools.enableRetry ?? true });

    const { valid, message } = isConfigValid(config);

    if (valid) {
      return config;
    } else {
      throw new Error(message);
    }
  } catch (e) {
    console.error(e);
    throw new Error('Commercetools - Mollie Integration configuration is missing or not provided in the valid JSON format');
  }
}

const ctMollieConfig = process.env.CT_MOLLIE_CONFIG;
const config = loadConfig(ctMollieConfig);

export default config;
