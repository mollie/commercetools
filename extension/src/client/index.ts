import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
