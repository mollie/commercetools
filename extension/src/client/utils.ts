import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
