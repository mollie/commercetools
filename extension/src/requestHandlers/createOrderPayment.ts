import { MollieClient, OrderPaymentCreateParams, Payment } from '@mollie/api-client';
import { formatMollieErrorResponse } from '../errorHandlers/formatMollieErrorResponse';
import { Action, CTTransactionType, CTUpdatesRequestedResponse } from '../types';
import { createDateNowString } from '../utils';
import Logger from '../logger/logger';

export function getOrdersPaymentsParams(ctObj: any): Promise<OrderPaymentCreateParams> {
  try {
    const parsedCreateOrderPaymentRequest = JSON.parse(ctObj?.custom?.fields?.createOrderPaymentRequest);
    const orderPaymentCreateParams = {
      orderId: ctObj?.key,
    };
    const { method, customerId, mandateId } = parsedCreateOrderPaymentRequest;
    if (method) Object.assign(orderPaymentCreateParams, { method });
    if (customerId) Object.assign(orderPaymentCreateParams, { customerId });
    if (mandateId) Object.assign(orderPaymentCreateParams, { mandateId });
    return Promise.resolve(orderPaymentCreateParams);
  } catch (error) {
    Logger.error({ error });
    return Promise.reject({ status: 400, title: 'Could not make parameters needed to create Mollie order payment.', field: 'createOrderPaymentRequest' });
  }
}

export function createCtActions(orderPaymentResponse: Payment, ctObj: any): Action[] {
  const stringifiedOrderPaymentResponse = JSON.stringify(orderPaymentResponse);
  const result: Action[] = [
    {
      action: 'addInterfaceInteraction',
      type: {
        key: 'ct-mollie-integration-interface-interaction-type',
      },
      fields: {
        actionType: 'createOrderPayment',
        createdAt: createDateNowString(),
        request: ctObj?.custom?.fields?.createOrderPaymentRequest,
        response: stringifiedOrderPaymentResponse,
      },
    },
    {
      action: 'setCustomField',
      name: 'createOrderPaymentResponse',
      value: stringifiedOrderPaymentResponse,
    },
    {
      action: 'addTransaction',
      transaction: {
        timestamp: orderPaymentResponse.createdAt,
        type: CTTransactionType.Charge,
        amount: ctObj.amountPlanned,
        interactionId: orderPaymentResponse.id,
      },
    },
  ];
  return result;
}

export default async function createOrderPayment(ctObj: any, mollieClient: MollieClient, getOrdersPaymentsParams: Function, createCtActions: Function): Promise<CTUpdatesRequestedResponse> {
  try {
    const ordersPaymentsParams = await getOrdersPaymentsParams(ctObj);
    const mollieOrderPaymentRes = await mollieClient.orders_payments.create(ordersPaymentsParams);
    const ctActions = createCtActions(mollieOrderPaymentRes, ctObj);
    return {
      actions: ctActions,
      status: 201,
    };
  } catch (error: any) {
    Logger.error({ error });
    const errorResponse = formatMollieErrorResponse(error);
    return errorResponse;
  }
}
