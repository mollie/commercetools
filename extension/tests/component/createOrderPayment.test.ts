import nock from 'nock';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import _ from 'lodash';
import { mocked } from 'ts-jest/utils';
import app from '../../src/app';
import config from '../../config/config';
import Logger from '../../src/logger/logger';
import { orderPaymentWithIDEAL, orderPaymentWithKlarna, orderHasOpenPaymentError } from './mockResponses/mollieData/createOrderPayment.data';
import { CTTransactionState, CTTransactionType } from '../../src/types';

jest.mock('uuid');

describe('Create Order Payment', () => {
  const {
    commercetools: { authUrl },
  } = config;
  const mockLogError = jest.fn();

  const paymentId = 'd75d0b1d-64c5-4c8f-86f6-b9510332e743';
  const mollieOrderId = 'ord_mzvyt8';
  const baseMockCTPaymentObj: any = {
    resource: {
      obj: {
        id: paymentId,
        key: mollieOrderId,
        paymentMethodInfo: {
          paymentInterface: 'mollie',
        },
        amountPlanned: {
          currencyCode: 'EUR',
          centAmount: 50000,
        },
        custom: {
          fields: {
            createPayment: '{}',
          },
        },
      },
    },
  };
  let authTokenScope: any;
  beforeAll(() => {
    // Ensure consistent uuid and datetime
    jest.spyOn(Date.prototype, 'toISOString').mockImplementation(() => '2021-11-10T14:02:45.858Z');
    mocked(uuid).mockReturnValue('b2bd1698-9923-4704-9729-02db2de495d1');
    // Credentials authentication flow is called first by commercetools client
    authTokenScope = nock(`${authUrl}`).persist().post('/oauth/token').reply(200, {
      access_token: 'vkFuQ6oTwj8_Ye4eiRSsqMeqLYNeQRJi',
      expires_in: 172800, // seconds (2 days)
      scope: 'manage_project:{projectKey}',
      token_type: 'Bearer',
    });
    // Prevent logs from cluttering test output
    Logger.error = mockLogError;
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    authTokenScope.persist(false);
  });

  describe('Happy Path', () => {
    afterEach(() => {
      nock.cleanAll();
    });

    it('Should return 201 when a new payment is created against the mollie order, changing from pay now (iDEAL) to pay later method (klarna)', async () => {
      const orderPaymentCreatedScope = nock('https://api.mollie.com/v2').post(`/orders/${mollieOrderId}/payments`, { method: 'klarnapaylater' }).reply(201, orderPaymentWithKlarna);

      const mockCTPaymentObj = _.cloneDeep(baseMockCTPaymentObj);
      mockCTPaymentObj.resource.obj.paymentMethodInfo.method = 'klarnapaylater';
      mockCTPaymentObj.resource.obj.transactions = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Charge,
          state: CTTransactionState.Failure,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Authorization,
          state: CTTransactionState.Initial,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, body } = res;
      expect(status).toBe(201);

      // Check actions
      const { actions } = body;
      expect(actions).toHaveLength(4);
      const updateStateAction = actions.find((action: any) => action.action === 'changeTransactionState');
      expect(updateStateAction.state).toBe('Pending');
      const updateIteractionIdAction = actions.find((action: any) => action.action === 'changeTransactionInteractionId');
      expect(updateIteractionIdAction.interactionId).toBe(orderPaymentWithKlarna.id);
      expect(updateStateAction.state).toBe('Pending');
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });

      expect(orderPaymentCreatedScope.isDone()).toBeTruthy();
    });

    it('Should return 201 when a new payment is created against the mollie order, changing from pay later (klarna) to pay now method (iDEAL)', async () => {
      const orderPaymentCreatedScope = nock('https://api.mollie.com/v2').post(`/orders/${mollieOrderId}/payments`, { method: 'ideal' }).reply(201, orderPaymentWithIDEAL);

      const mockCTPaymentObj = _.cloneDeep(baseMockCTPaymentObj);
      mockCTPaymentObj.resource.obj.paymentMethodInfo.method = 'ideal';
      mockCTPaymentObj.resource.obj.transactions = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Authorization,
          state: CTTransactionState.Failure,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Charge,
          state: CTTransactionState.Initial,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, body } = res;
      expect(status).toBe(201);

      // Check actions
      const { actions } = body;
      expect(actions).toHaveLength(4);
      const updateStateAction = actions.find((action: any) => action.action === 'changeTransactionState');
      expect(updateStateAction.state).toBe('Pending');
      const updateIteractionIdAction = actions.find((action: any) => action.action === 'changeTransactionInteractionId');
      expect(updateIteractionIdAction.interactionId).toBe(orderPaymentWithIDEAL.id);
      expect(updateStateAction.state).toBe('Pending');
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });

      expect(orderPaymentCreatedScope.isDone()).toBeTruthy();
    });

    it('Should return 201 when a new payment is created against the mollie order, changing payment method from klarnasliceit to klarnapaylater', async () => {
      const orderPaymentCreatedScope = nock('https://api.mollie.com/v2').post(`/orders/${mollieOrderId}/payments`, { method: 'klarnapaylater' }).reply(201, orderPaymentWithKlarna);

      const mockCTPaymentObj = _.cloneDeep(baseMockCTPaymentObj);
      mockCTPaymentObj.resource.obj.paymentMethodInfo.method = 'klarnapaylater';
      mockCTPaymentObj.resource.obj.transactions = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Authorization,
          state: CTTransactionState.Failure,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Authorization,
          state: CTTransactionState.Initial,
          amount: {
            currencyCode: 'EUR',
            centAmount: 90000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);
      const { status, body } = res;
      expect(status).toBe(201);

      // Check actions
      const { actions } = body;
      expect(actions).toHaveLength(4);
      const updateStateAction = actions.find((action: any) => action.action === 'changeTransactionState');
      expect(updateStateAction.state).toBe('Pending');
      const updateIteractionIdAction = actions.find((action: any) => action.action === 'changeTransactionInteractionId');
      expect(updateIteractionIdAction.interactionId).toBe(orderPaymentWithKlarna.id);
      expect(updateStateAction.state).toBe('Pending');
      actions.forEach((action: any) => {
        expect(action).toMatchSnapshot();
      });

      expect(orderPaymentCreatedScope.isDone()).toBeTruthy();
    });
  });

  describe('Unhappy Path', () => {
    it('Should return 400 when the mollie order has an open payment', async () => {
      const orderWithOpenPaymentScope = nock('https://api.mollie.com/v2').post(`/orders/${mollieOrderId}/payments`, { method: 'klarnapaylater' }).reply(422, orderHasOpenPaymentError);

      const mockCTPaymentObj = _.cloneDeep(baseMockCTPaymentObj);
      mockCTPaymentObj.resource.obj.paymentMethodInfo.method = 'klarnapaylater';
      mockCTPaymentObj.resource.obj.transactions = [
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Charge,
          state: CTTransactionState.Pending,
          amount: {
            currencyCode: 'EUR',
            centAmount: 50000,
          },
        },
        {
          id: '2b5f68ad-ae94-4bf1-ae41-7096e5142f89',
          type: CTTransactionType.Authorization,
          state: CTTransactionState.Initial,
          amount: {
            currencyCode: 'EUR',
            centAmount: 50000,
          },
        },
      ];

      const res = await request(app).post('/').send(mockCTPaymentObj);

      const { status, body } = res;
      expect(status).toBe(400);

      const { errors } = body;
      expect(errors).toHaveLength(1);
      expect(errors[0]).toEqual({
        code: 'SemanticError',
        message: 'Cannot create a new payment for order ord_byxxzq when it has an open payment.',
        extensionExtraInfo: {
          originalStatusCode: 422,
          title: 'Unprocessable Entity',
          links: {
            documentation: {
              href: 'https://docs.mollie.com/overview/handling-errors',
              type: 'text/html',
            },
          },
        },
      });
      expect(orderWithOpenPaymentScope.isDone()).toBeTruthy();
    });
  });
});
