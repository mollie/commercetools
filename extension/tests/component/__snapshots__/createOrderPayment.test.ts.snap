// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay later (klarna) to pay now method (iDEAL) 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrderPayment",
    "createdAt": "2022-01-17T13:18:53+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"paymentMethod\\":\\"ideal\\"}",
    "response": "{\\"molliePaymentId\\":\\"tr_DCNv63fDhy\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/test-mode?method=klarnapaylater&token=3.ld7bf8\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay later (klarna) to pay now method (iDEAL) 2`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_DCNv63fDhy",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay later (klarna) to pay now method (iDEAL) 3`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2022-01-17T13:18:53+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay later (klarna) to pay now method (iDEAL) 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay now (iDEAL) to pay later method (klarna) 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrderPayment",
    "createdAt": "2022-01-17T13:18:53+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"paymentMethod\\":\\"klarnapaylater\\"}",
    "response": "{\\"molliePaymentId\\":\\"tr_DCNv63fDhy\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/test-mode?method=klarnapaylater&token=3.ld7bf8\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay now (iDEAL) to pay later method (klarna) 2`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_DCNv63fDhy",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay now (iDEAL) to pay later method (klarna) 3`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2022-01-17T13:18:53+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing from pay now (iDEAL) to pay later method (klarna) 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing payment method from klarnasliceit to klarnapaylater 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "createOrderPayment",
    "createdAt": "2022-01-17T13:18:53+00:00",
    "id": "b2bd1698-9923-4704-9729-02db2de495d1",
    "request": "{\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\",\\"paymentMethod\\":\\"klarnapaylater\\"}",
    "response": "{\\"molliePaymentId\\":\\"tr_DCNv63fDhy\\",\\"checkoutUrl\\":\\"https://www.mollie.com/checkout/test-mode?method=klarnapaylater&token=3.ld7bf8\\",\\"transactionId\\":\\"2b5f68ad-ae94-4bf1-ae41-7096e5142f89\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing payment method from klarnasliceit to klarnapaylater 2`] = `
Object {
  "action": "changeTransactionInteractionId",
  "interactionId": "tr_DCNv63fDhy",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing payment method from klarnasliceit to klarnapaylater 3`] = `
Object {
  "action": "changeTransactionTimestamp",
  "timestamp": "2022-01-17T13:18:53+00:00",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;

exports[`Create Order Payment Happy Path Should return 201 when a new payment is created against the mollie order, changing payment method from klarnasliceit to klarnapaylater 4`] = `
Object {
  "action": "changeTransactionState",
  "state": "Pending",
  "transactionId": "2b5f68ad-ae94-4bf1-ae41-7096e5142f89",
}
`;
